### ISSUE TITLE
New value type (null) is not in annotated types - need type inference through ternary

### ISSUE THREAD

**Issue by iakoder** [kalessil/phpinspectionsea#1119]
| Subject        | Details                                                                       |
| :------------- | :---------------------------------------------------------------------------- |
| Issue type     | False-Positive / Feature request / Support request                          |
| Plugin         | Php Inspections (EA Extended), 3.0.8.1 |
| Language level | PHP 7.0                        |

The $p1 != '' conditional guarantees that null will not be assigned into $p2 but PHP Inspections doesn't seem to agree. 

```
<?php

/**
 * @param bool|float|int|null|string  $p1
 * @param bool|float|int|string       $p2
 */
function f($p1, &$p2) {
    // New value type (null) is not in annotated types.
    $p2 = ( $p1 != '' ? $p1 : '');
}
```

 I've thought of two workarounds:
- @noinspection
-  $p2 = ( $p1 != '' ? $p1 : '') ?? '';

Is this anything that the plugin would ever be able to detect?

PhpStorm 2018.3
Build #PS-183.4284.150, built on November 21, 2018
Licensed to Intacct / Dan Wilks
Subscription is active until June 23, 2019
JRE: 1.8.0_152-release-1343-b15 x86_64
JVM: OpenJDK 64-Bit Server VM by JetBrains s.r.o
macOS 10.14.1

**Comment by funivan** [comment]
> The $p1 != '' conditional guarantees that null will not be assigned into $p2 but PHP Inspections doesn't seem to agree.
This condition guarantees that not empty string will not be assigned into $p2
Try to run this code when p1 is null

**Comment by iakoder** [comment]
php -r '$p1=null; $p2 = ( $p1 != "" ? $p1 : ""); var_dump($p2);'
Command line code:1:
string(0) ""

**Comment by funivan** [comment]
@iakoder you are right. This is not a strict comparison so we should filter null values here.

**Comment by kalessil** [comment]
`$p2 = $p1 ?: '';` shoulddo the trick. The original case types are resolved by ide, while the short one is handled on our end.

**Comment by iakoder** [comment]
@kalessil Unfortunately $p1 ?: '' is not the same as $p1 != "" ? $p1 : "" since the first case converts "0" to "" whereas the second does not.

**Comment by kalessil** [comment]
Oh, my bad. Semantics is really differs.

**Comment by funivan** [comment]
@kalessil do you plan to fix it?

**Comment by kalessil** [comment]
Nope, it's not realistic IMO to produce reliable enough fix here.

**Comment by funivan** [comment]
Agree. 
To prevent this behavior it is better to use strict type comparison + avoid union types.  IMHO
